src/pyhf/contrib/utils.py

```diff
@@ -62,17 +66,64 @@ def download(archive_url, output_directory, force=False, compress=False):
         # The HEPData landing page for the resource file can check if the Accept
         # request HTTP header matches the content type of the resource file and
         # return the content directly if so.
+        # TODO: Figure out how to accept both x-tar and zip headers. :?
+        # with requests.get(
+        #     archive_url, headers={"Accept": "application/zip"}
+        # ) as response:
         with requests.get(
             archive_url, headers={"Accept": "application/x-tar"}
         ) as response:
+            if response.status_code != 200:
+                raise exceptions.InvalidArchive(
+                    f"{archive_url} gives a response code of {response.status_code}.\n"
+                    + "There is either something temporarily wrong with the archive host"
+                    + f" or {archive_url} is an invalid URL."
+                )
+
             if compress:
                 with open(output_directory, "wb") as archive:
                     archive.write(response.content)
             else:
-                with tarfile.open(
-                    mode="r|gz", fileobj=BytesIO(response.content)
-                ) as archive:
-                    archive.extractall(output_directory)
+                # Support for file-like objects for tarfile.is_tarfile was added
+                # in Python 3.9, so as pyhf is currently Python 3.7+ then can't
+                # do tarfile.is_tarfile(BytesIO(response.content)).
+                # Instead, just use a 'try except' block to determine if the
+                # archive is a valid tarfile.
+                # TODO: Simplify after pyhf is Python 3.9+ only
+                try:
+                    # Use transparent compression to allow for .tar or .tar.gz
+                    with tarfile.open(
+                        mode="r:*", fileobj=BytesIO(response.content)
+                    ) as archive:
+                        archive.extractall(output_directory)
+                except tarfile.ReadError:
+                    if not zipfile.is_zipfile(BytesIO(response.content)):
+                        raise exceptions.InvalidArchive(
+                            f"The archive downloaded from {archive_url} is not a tarfile"
+                            + " or a zipfile and so can not be opened as one."
+                        )
```

Comment:
@kratsg as you have a lot more experience mocking out things in the codebase then I do, do you have any thoughts on how to best approach mocking the requests `response.content` as binary representations of a tarfile and zipfile?

Reply:
You could mock all the way through the calls I think: mock `tarfile.open` to succeed / fail, `BytesIO` to return anything you'd like, and mock `zipfile.is_zipfile` to return `False` given whatever return value you pick from `BytesIO`. That would assume though that those APIs do not break, which is probably safe enough for the standard library. That way it would not matter what `response.content` is either, that could be a mock itself too.